x-pack/plugins/triggers_actions_ui/public/application/sections/alerts_table/hooks/use_columns.ts

```diff
@@ -0,0 +1,228 @@
+  if (fieldType === 'object') return 'json';
+  return fieldType;
+};
+
+/**
+ * EUI Data Grid expects the columns to have a property 'schema' defined for proper sorting
+ * this schema as its own types as can be check out in the docs so we add it here manually
+ * https://eui.elastic.co/#/tabular-content/data-grid-schema-columns
+ */
+const euiColumnFactory = (
+  column: EuiDataGridColumn,
+  browserFields: BrowserFields
+): EuiDataGridColumn => {
+  const browserFieldsProps = getBrowserFieldProps(column.id, browserFields);
+  return {
+    ...column,
+    schema: fieldTypeToDataGridColumnTypeMapper(browserFieldsProps.type),
+  };
+};
+
+/**
+ * Searches in browser fields object for a specific field
+ */
+const getBrowserFieldProps = (
+  _columnId: string,
+  browserFields: BrowserFields
+): Partial<BrowserField> => {
+  for (const [, categoryDescriptor] of Object.entries(browserFields)) {
+    if (!categoryDescriptor.fields) {
+      continue;
+    }
+
+    for (const [fieldName, fieldDescriptor] of Object.entries(categoryDescriptor.fields)) {
+      if (fieldName === _columnId) {
+        return fieldDescriptor;
+      }
+    }
+  }
+  return { type: 'string' };
+};
+
+/**
+ * @param columns Columns to be considered in the alerts table
+ * @param browserFields constant object with all field capabilities
+ * @returns columns but with the info needed by the data grid to work as expected, e.g sorting
+ */
+const populateColumns = (
+  columns: EuiDataGridColumn[],
+  browserFields: BrowserFields
+): EuiDataGridColumn[] => {
+  return columns.map((column: EuiDataGridColumn) => {
+    return euiColumnFactory(column, browserFields);
+  });
+};
+
+const getColumnByColumnId = (columns: EuiDataGridColumn[], columnId: string) => {
+  return columns.find(({ id }: { id: string }) => id === columnId);
+};
+
+const persist = ({
+  id,
+  storageAlertsTable,
+  columns,
+  visibleColumns,
+  storage,
+}: {
+  id: string;
+  storageAlertsTable: React.MutableRefObject<AlertsTableStorage>;
+  storage: React.MutableRefObject<IStorageWrapper>;
+  columns: EuiDataGridColumn[];
+  visibleColumns: string[];
+}) => {
+  storageAlertsTable.current = {
+    ...storageAlertsTable.current,
+    columns,
+    visibleColumns,
+  };
+  storage.current.set(id, storageAlertsTable.current);
+};
+
+export const useColumns = ({
+  featureIds,
+  storageAlertsTable,
+  storage,
+  id,
+  defaultColumns,
+}: UseColumnsArgs) => {
+  const [isBrowserFieldDataLoading, browserFields] = useFetchBrowserFieldCapabilities({
+    featureIds,
+  });
+  const [columns, setColumns] = useState<EuiDataGridColumn[]>(storageAlertsTable.current.columns);
+  const [isColumnsPopulated, setColumnsPopulated] = useState<boolean>(false);
+  const [visibleColumns, setVisibleColumns] = useState(
+    storageAlertsTable.current.visibleColumns ?? []
+  );
+
+  useEffect(() => {
+    if (isBrowserFieldDataLoading !== false || isColumnsPopulated === true) return;
```

Comment:
why not `if (isBrowserFieldDataLoading || isColumnsPopulated) return`? they are both booleans.